### ISSUE TITLE
Add find_cycle functions

### ISSUE THREAD

**Issue by mtreinish** [Qiskit/retworkx#155]
### What is the expected enhancement?

Retworkx is missing functionality equivalent to networkx's `find_cycle` method: https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.cycles.find_cycle.html
while retworkx does have `cycle_basis` functions it doesn't have a method to find cycles from arbitrary nodes. However, to start the retworkx functions added don't need to have the same level of flexibility as networkx's. The immediate use case for this is: https://github.com/Qiskit/qiskit-terra/blob/779fa14d4452e35306ca849d98e56f8aa40d14be/qiskit/transpiler/passes/routing/algorithms/token_swapper.py#L160 which only cares about finding a cycle from a source node without an orientation. If the need arises after that is added we can always expand it later.

**Comment by darknight009** [comment]
Hi! If no one is working, I'd like to take this up.

**Comment by mtreinish** [comment]
@darknight009 sure, feel free to give it a go. I've assigned the issue to you. Let me know if you need any help or have any questions.

**Comment by darknight009** [comment]
Thank you! What kind arguments are expected? Would they be same as cycle_basis function where PyGraph is expected with an optional node?

**Comment by mtreinish** [comment]
Yeah we probably should implement this for a `PyDiGraph` to start, although depending on the implementation it might not be hard to add 2 functions `graph_find_cycle` and `digraph_find_cycle` for both a directed and undirected graph.

But, I would say the minim we probably need would be a function signature like:

```rust
pub fn digraph_find_cycle(graph: digraph::PyDiGraph, source: usize) -> PyResult<Vec<(usize, usize)>>
```
or some variant like that. But I'm guessing a lot of this will depend on the implementation of the algorithm so don't feel locked into that.

I'm just thinking at least to start we just need coverage for what qiskit-terra is doing now, so on a directed graph, always specifying a single source node and using the default `None` orientation. But, as long as the terra use case is covered and we return the same results as terra expects we have some flexibility on the exact signature.

**Comment by darknight009** [comment]
Hi @mtreinish! I have added the function and test cases but I'm unclear how to update the docs.

**Comment by mtreinish** [comment]
> I have added the function and test cases but I'm unclear how to update the docs.

Ah, sorry I probably should add better documentation on how the documentation is put together. In this case to add the docs for the new function all you need to do is add the new function to the index in `docs/source/api.rst` under the Algorithm Functions section: https://github.com/Qiskit/retworkx/blob/master/docs/source/api.rst#algorithm-functions once you add the function there sphinx will automatically build the docs for the function from the docstring you wrote on the function.

You can test the docs builds locally with `tox -edocs` which will build the docs as an html page you can locally view in your browser from `docs/build/html/index.html` (https://github.com/Qiskit/retworkx/blob/master/CONTRIBUTING.md#building-documentation)
